### ISSUE TITLE
Partial query data is reused when `notifyOnNetworkStatusChange: true` and a nested field has an argument

### ISSUE THREAD

**Issue by amannn** [apollographql/apollo-client#8715]
After upgrading from `@apollo/client@3.4.7` to `3.4.8` I found that a query in the code base I'm working on suddenly reuses a partial result from a different query when `notifyOnNetworkStatusChange` is set to `true` and causes the app to crash. Please see the reproduction below.

**Intended outcome:**

Apollo Client should never reuse partial results from a different query when `returnPartialData` is `false`.

**Actual outcome:**

Apollo Client returns partial results temporarily which causes the app to crash.

**How to reproduce the issue:**

See this CodeSandbox: https://codesandbox.io/s/apollo-client-error-reproduction-partial-data-dw5sc

Note that once you remove `notifyOnNetworkStatusChange`, the bug isn't there anymore. The other aspect that needs to be true, is that the nested object (in the example the field `address`) needs an argument – if you remove the argument from the query, the bug also doesn't occur.

**Versions**

`@apollo/client@3.4.9`

Thank you for your help!

**Comment by brainkim** [comment]
I’ve spent a good day investigating this now, and not really getting the gist of it. It definitely seems to be related to changes in https://github.com/apollographql/apollo-client/pull/8642. ObservableQuery is saving partial results to `this.last` for some reason, and it’s leaking into the result. It’s also interesting that this is only happening when a query has a default variable which is used because a variable is not provided to the query. Will require a good deep investigation. 🙈

**Comment by amannn** [comment]
Thanks a lot for looking into this 👏

**Comment by ozmoroz** [comment]
Bikeexchange reported the same issue on https://www.cycling-emporium.com

To reproduce:
1. Go to /s
1. Click on a brand from the main nav
1. The brand (/shop) page should load
1. Click on a different brand from the main nav
1. The page should fail to load and only skeleton UI should be seen.

**Comment by hueter** [comment]
Thank you @amannn for reporting this issue, I have been facing the same thing and was about to open a new issue when I finally found you had already done so! I forked your sandbox (very helpful btw 👍 ) and modified things to be almost my exact production use case: https://codesandbox.io/s/apollo-client-error-reproduction-missing-cache-result-fields-le217?file=/src/App.js 

I left a bunch of comments and unnecessary styling 😅  but TL;DR if you click the big red button, it triggers a refetch that produces the bug.

I also discovered during debugging that there was a "Missing cache result fields" error that I got to show up in the console by [setting the log verbosity](https://github.com/apollographql/apollo-client/pull/8489).

This has been blocking my team from upgrading past 3.4.7 sadly; but I have only just had time to investigate it recently and also concluded that #8642 was the most likely culprit here.

Removing `notifyOnNetworkStatusChange: true` also appears to resolve the issue for me, but that breaks the "smooth" loading behavior in the tables across my app (for things like pagination and filtering).

Happy to provide more context if necessary. Luckily I have a Cypress spec on CI that fails whenever I try a new Apollo version where it's still affected so I will also report back if it coincidentally gets fixed somehow lol.
